p2pfl/learning/aggregators/scaffold.py

```diff
@@ -0,0 +1,120 @@
+    """
+
+    def __init__(self, node_name:str, **kwargs):
+        """
+        Initialize the aggregator.
+
+        Args:
+            node_name: The name of the node.
+            kwargs: Additional arguments.
+
+        """
+        super().__init__(node_name)
+        self.c = [] # global control variates
+        self.global_lr = kwargs.get('global_lr', 0.1)
+
+    def aggregate(self, models: List[P2PFLModel]) -> P2PFLModel:
+        """
+        Aggregate the models and control variates from clients.
+
+        Args:
+            models: List of models to aggregate.
+
+        """
+        if not models:
+            raise NoModelsToAggregateError(f"({self.node_name}) Trying to aggregate models when there is no models")
+        total_samples = sum([m.get_num_samples() for m in models])
+        # initialize the accumulators for the model and the control variates
+        first_model_weights = models[0].get_parameters()
+        accum_y = [np.zeros_like(layer) for layer in first_model_weights]
+
+        # Accumulate weighted model updates
+        for m in models:
+            delta_y_i = m.get_info('delta_y_i')
+
+
+            if delta_y_i is None:
+                raise ValueError(f"Model from node '{self.node_name}' is missing 'delta_y_i'.")
+            num_samples = m.get_num_samples()
+            for i, layer in enumerate(delta_y_i):
+                accum_y[i] += layer * num_samples
+
+        # Normalize the accumulated model updates
+        accum_y = [layer / total_samples for layer in accum_y]
+        accum_y = [layer * self.global_lr for layer in accum_y] # apply global learning rate
+
+        # Accumulate control variates
+        accum_c = None
+        for m in models:
+            delta_c_i = m.get_info('delta_c_i')
+            if delta_c_i is None:
+                raise ValueError("Node is missing 'delta_c_i'.")
+            if accum_c is None:
+                accum_c = [layer.copy() for layer in delta_c_i]
+            else:
+                for i in range(len(accum_c)):
+                    accum_c[i] += delta_c_i[i]
+
+        # Normalize the accumulated control variates
+        num_models = len(models)
+        accum_c = [layer / num_models for layer in accum_c]
+
+        # Update global c
+        if not self.c:
+            self.c = [np.zeros_like(layer) for layer in accum_c]
+
+        for i in range(len(self.c)):
+            self.c[i] += accum_c[i]
+
+        # Get contributors
+        contributors = []
+        for m in models:
+            contributors.extend(m.get_contributors())
+
+        aggregated_model = models[0].build_copy(
+            params=accum_y,
+            num_samples=total_samples,
+            contributors=contributors
+        )
+        aggregated_model.add_info('global_c', self.c)
+        return aggregated_model
+
+    def get_required_callbacks(self) -> List[str]:
```

Comment:
In the same way that you have a get_required_callbacks, a get_required_additional_info would be great. And it should include a validation of the additional_info parameters for security reasons